src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Interop/JavaScriptExports.cs

```diff
@@ -140,37 +151,23 @@ public static void ReleaseJSOwnedObjectByGCHandle(JSMarshalerArgument* arguments
         {
             ref JSMarshalerArgument arg_exc = ref arguments_buffer[0]; // initialized by caller in alloc_stack_frame()
             ref JSMarshalerArgument arg_1 = ref arguments_buffer[2]; // initialized and set by caller
+
             try
             {
-                var gcHandle = arg_1.slot.GCHandle;
-                if (IsGCVHandle(gcHandle))
-                {
-                    if (ThreadJsOwnedHolders.Remove(gcHandle, out PromiseHolder? holder))
-                    {
-                        holder.GCHandle = IntPtr.Zero;
-                        holder.Callback!(null);
-                    }
-                }
-                else
-                {
-                    GCHandle handle = (GCHandle)gcHandle;
-                    var target = handle.Target!;
-                    if (target is PromiseHolder holder)
-                    {
-                        holder.GCHandle = IntPtr.Zero;
-                        holder.Callback!(null);
-                    }
-                    else
-                    {
-                        ThreadJsOwnedObjects.Remove(target);
-                    }
-                    handle.Free();
-                }
+                // when we arrive here, we assume that all proxies are owned by calling thread
+                var ctx = JSProxyContext.CapturedInstance = JSProxyContext.AssertCurrentContext();
+                ctx.ReleaseJSOwnedObjectByGCHandle(arg_1.slot.GCHandle);
             }
             catch (Exception ex)
             {
                 arg_exc.ToJS(ex);
             }
+#if FEATURE_WASM_THREADS
+            finally
+            {
+                JSProxyContext.CapturedInstance = null;
```

Comment:
Are we certain by construction that when we null this out, it is always going to currently be `== ctx`? Or is it possible something called by our wrapper could change it to another value, and now we'd be messing up state?
Should we be recording what it was at entry and restoring to that instead of setting it to `null`? This feels wrong to me.

Maybe the shape should be like
`JSProxyContext.PushCurrentContext(JSProxyContext.AssertCurrentContext())` paired with `JSProxyContext.PopCurrentContext()`

Reply:
If we go with push/pop i think the ideal would be
```csharp
var ctx = JSProxyContext.CaptureCurrentContext();
...
JSProxyContext.PopCurrentContext(ctx);
```